sourcesapi/sourcebaseclass.py

```diff
@@ -0,0 +1,236 @@
+            1 -- connected refused
+            2 -- api key
+            3 -- rate limit
+            4 -- timeout
+    """
+
+    @staticmethod
+    def default_msg(sub_code):
+        if sub_code == 1:
+            return "The connection was refused."
+        if sub_code == 2:
+            return "An API key error occurred."
+        if sub_code == 3:
+            return "Rate limit exceeded. Please wait a bit or enter a personal key if using the default."
+        if sub_code == 4:
+            return "The connection timed out."
+        if sub_code == 5:
+            return "Number of retries exceeded."
+
+        return "General network error."
+
+    def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
+        if desc == "":
+            desc = self.default_msg(sub_code)
+
+        super().__init__(source, 2, desc, sub_code)
+
+
+class SourceInfoDataError(SourceInfoError):
+    """Data class exception for information sources
+
+    Attributes:
+        sub_code -- numerical code for finer detail
+            1 -- unexpected data
+            2 -- malformed data
+            3 -- missing data
+    """
+
+    @staticmethod
+    def default_msg(sub_code):
+        if sub_code == 1:
+            return "Unexpected data encountered."
+        if sub_code == 2:
+            return "Malformed data encountered."
+        if sub_code == 3:
+            return "Missing data encountered."
+
+    def __init__(self, source: str = "", sub_code: int = 0, desc: str = "") -> None:
+        if desc == "":
+            desc = self.default_msg(sub_code)
+
+        super().__init__(source, 3, desc, sub_code)
+
+
+# Class talkers instance
+class InfoSource:
+    """This is the class for mysource."""
+
+    def __init__(self) -> None:
+        # Identity name for the information source etc.
+        self.source_details: SourceDetails | None = None  # Can use this to test if custom talker has been configured
+
+        self.log_func: Callable[[str], None] | None = None
+        self.logger = logger
+
+    def set_log_func(self, log_func: Callable[[str], None]) -> None:
+        self.log_func = log_func
+
+    # issueidentifier will set_log_func to print any write_log to console otherwise logger.info is not printed
+    def write_log(self, text: str) -> None:
+        if self.log_func is None:
+            self.logger.info(text)
+        else:
+            self.log_func(text)
+
+    def check_api_key(self, key: str, url: str):
+        raise NotImplementedError
+
+    # Search for series/volumes
+    def search_for_series(
+        self,
+        series_name: str,
+        callback: Callable[[int, int], None] | None = None,
+        refresh_cache: bool = False,
+        literal: bool = False,
+    ) -> list[ComicSearchResults]:
+        raise NotImplementedError
+
+    # Get volume/series information
+    def fetch_volume_data(self, series_id: int, settings: ComicTaggerSettings) -> GenericMetadata:
```

Comment:
These definitions are not up to date with your usage in sourcesapi.py

Reply:
Do you mean the "settings" being passed along? That was due to the async stuff iirc. I left a note in the init when loading the ComicTaggerSettings. "
```
43: # No easy way around this due to QT network and async with coverimagewidget
44: self.settings = ComicTaggerSettings("")
```
I think it was causing a circular import. Altering coverimagewidget to not use async might resolve that problem. Or, is the problem something else I've missed?

Reply:
I was looking at `SourcesApi` vs `ComicVineTalker` but you still don't need to pass the entire settings object through, the only place you use it is in `map_cv_data_to_metadata` which uses exactly one attribute from the settings

Reply:
Also why is `fetch_volume_data` not implemented on `ComicVineTalker`?

Reply:
> I was looking at `SourcesApi` vs `ComicVineTalker` but you still don't need to pass the entire settings object through, the only place you use it is in `map_cv_data_to_metadata` which uses exactly one attribute from the settings

I was following what was already there with the settings. The other issue is that it should (for the future) pass all the comic source general settings through but as all the settings are monolithic, passing the whole object it was. (This could change with the new settings rework?) This does bring up the question of if everything in the comic source general tab should be handled outside of the talker? The API should make the changes for use_series_start_as_volume as the example here?

Reply:
> Also why is `fetch_volume_data` not implemented on `ComicVineTalker`?

When testing with just the volume data from CV, I renamed the "fetch_volume_data" to "fetch_partial_volume_data" (it's only called internally to CV Talker) as that's what it does and to keep the other inline with "fetch_issue_data" naming. I've left it in for the future with sources that only have volume data.

Reply:
> When testing with just the volume data from CV, I renamed the "fetch_volume_data" to "fetch_partial_volume_data" (it's only called internally to CV Talker) as that's what it does and to keep the other inline with "fetch_issue_data" naming. I've left it in for the future with sources that only have volume data.

Then let's either leave it out entirely or implement it

> I was following what was already there with the settings. The other issue is that it should (for the future) pass all the comic source general settings through but as all the settings are monolithic, passing the whole object it was. (This could change with the new settings rework?) This does bring up the question of if everything in the comic source general tab should be handled outside of the talker? The API should make the changes for use_series_start_as_volume as the example here?

So what you are doing with this pull request is you are making a completely separate package that comictaggerlib will use in order to retrieve information about comics, you can even see this in the imports.

sources/comicvine.py:
```python
from comictaggerlib import ctversion                     # Literally a version number
from comictaggerlib.comiccacher import ComicCacher       # This should actually be moved into this package
from comictaggerlib.settings import ComicTaggerSettings  # Settings storage
```
resulttypes.py: None
sourcebaseclass.py:
```python
from comictaggerlib.settings import ComicTaggerSettings  # Only used for typing no functional impact
```
sourcesapi.py:
```python
from comictaggerlib.settings import ComicTaggerSettings  # You instantiate the settings, which is actually wrong comictaggerlib already does this
```
sourcesutils.py: None

When a library is extracted from a program it needs to be generalized, even more so since this is intended to make it so that adding another source will be simple, which means that the person extracting the library needs to know exactly what goes in and what goes out, if everything is passed in then any change can break it because the library states that it needs everything. When a function/class uses the `ComicTaggerSettings` object it says that any change that happens to the `ComicTaggerSettings` object may break it, which is fine inside of `comictaggerlib` because it only exists as a whole program, SourcesAPI is now a separate package and it shouldn't break if I decide to remove the `last_main_window_height` attribute from `ComicTaggerSettings`.
You created a detailed way to record settings in `static_options` and `settings_options` you even have an option `remove_html_tables` which is used on the line directly above the only use of `ComicTaggerSettings`, this is more about consistency rather than where things should be handled as you have already started handling them with `static_options` and `settings_options`.

Reply:
I think the answer to this one is simple: Any option that is required by the sourceapi should not live in the "general" settings. I was unsure about where to put the use_series_start_as_volume setting but it makes sense it should be a per source option. The other two options are handled outside of the sourceapi.

Reply:
Pick whichever you want for now, I think there will be some bigger changes to the API by the time this gets merged but I want to at least resolve the major merge conflicts before any are even considered.